File: algorithms/rabin-karp/rabin-karp.py
def rabin_karp(text, pattern):
    """Returns index of a match of -1
    Rabin Karp algorithm for pattern matching
    time O(n+m) space O(1)
      >>> rabin_karp('', 'aab')
      -1
      >>> rabin_karp('abc', '')
      -1
      >>> rabin_karp('aabaacaa', 'aaa')
      -1
      >>> rabin_karp('a', 'aa')
      -1
      >>> rabin_karp('aa', 'a')
      0
      >>> rabin_karp('abcabc', 'abc')
      0
      >>> rabin_karp('aaabbcabc', 'abc')
      6
      >>> rabin_karp('aaaaaaaabaab', 'aab')
      6
    """
    n = len(text)
    m = len(pattern)
    if not n or not m or m > n:
        return -1
    BASE = 256
    PRIME = 101
    reset_hash = 1
    text_hash = 0
    pattern_hash = 0

    for i in range(m-1):
        reset_hash = (reset_hash * BASE) % PRIME

    for i in range(m):
        pattern_hash = (pattern_hash * BASE + ord(pattern[i])) % PRIME
        text_hash = (text_hash * BASE + ord(text[i])) % PRIME

    for i in range(n-m+1):
        if text_hash == pattern_hash:
            for j in range(m):
                if text[i+j] != pattern[j]:
                    break
                elif j+1 == m:
                    return i

        if i < n-m:
            text_hash = (
                BASE * (text_hash - ord(text[i]) * reset_hash) + ord(text[i+m])) % PRIME

        if text_hash < 0:
            text_hash += PRIME

    return -1


if __name__ == "__main__":
    import doctest
    doctest.testmod()

File: algorithms/rabin-karp/rabin-karp.js
/**
 * Rabin Karp algorithm
 * O(n) pattern matching algorithm with rolling hash
 * @param {String} text
 * @param {String} pattern
 * @returns {Number} index of match or -1
 */
function rabinKarp(text, pattern) {
  if (!text || !pattern || text.length < pattern.length) return -1;

  const PRIME = 101;
  const BASE = 256;
  const n = text.length;
  const m = pattern.length;
  let hashReset = 1;
  let textHash = 0;
  let patternHash = 0;

  // hashReset == BASE**(m-1) % PRIME
  for (let i = 0; i < m - 1; i++) {
    hashReset = (hashReset * BASE) % PRIME;
  }

  for (let i = 0; i < m; i++) {
    patternHash = (patternHash * BASE + pattern.charCodeAt(i)) % PRIME;
    textHash = (textHash * BASE + text.charCodeAt(i)) % PRIME;
  }

  for (let i = 0; i <= n - m; i++) {
    if (textHash === patternHash) {
      inner: for (let j = i; j < m + i; ) {
        if (text[j] !== pattern[j - i]) break inner;
        j++;
        if (j === m + i) return i;
      }
    }
    // recalculate rolling hash except for last iteration
    if (i < n - m) {
      textHash =
        (BASE * (textHash - text.charCodeAt(i) * hashReset) + // remove leading char
          text.charCodeAt(i + m)) % // add next char
        PRIME;

      if (textHash < 0) textHash += PRIME;
    }
  }
  return -1;
}

console.assert(rabinKarp('', 'aab') == -1, "rabinKarp('', 'aab') == -1");
console.assert(rabinKarp('abc', '') == -1, "rabinKarp('abc', '') == -1");
console.assert(rabinKarp('aabaacaa', 'aaa') == -1,"rabinKarp('aabaacaa', 'aaa') == -1");
console.assert(rabinKarp('a', 'aa') == -1, "rabinKarp('a', 'aa') == -1");
console.assert(rabinKarp('aa', 'a') == 0, "rabinKarp('aa', 'a') == 0");
console.assert(rabinKarp('abcabc', 'abc') == 0,"rabinKarp('abcccc', 'abc') == 0");
console.assert(rabinKarp('aaabbcabc', 'abc') == 6,"rabinKarp('aaabbcabc', 'abc') == 6");
console.assert(rabinKarp('aaaaaaaabaab', 'aab') == 6,"rabinKarp('aaaaaaaab', 'aab') == 6");
